File: ITB2/src/task2/GammaFilter.java
package task2;

import itb2.filter.AbstractFilter;
import itb2.image.Image;
import itb2.image.ImageFactory;

import static java.lang.Math.round;

public class GammaFilter extends AbstractFilter {
    private static final String GAMMA = "gamma";

    public GammaFilter(){
        this.properties.addDoubleProperty("gamma", 1);
    }

    private double calcGammaCorrection(double x, double gamma){

        x = round(256 * Math.pow(x/255, gamma));
        return ((x == 256) ? 255 : x);
    }

    @Override
    public Image filter(Image input) {

        Image output = ImageFactory.bytePrecision().gray(input.getSize());
        double gamma = properties.getDoubleProperty(GAMMA);
        for (int col = 0; col < input.getWidth(); col++){
            for (int row = 0; row < input.getHeight(); row++){
                double sum = 0;
                for(int chan = 0; chan < input.getChannelCount(); chan++)
                    sum += input.getValue(col, row, chan);
                output.setValue(col, row, calcGammaCorrection(sum/input.getChannelCount(), gamma));
            }
        }
        return output;
    }
}

File: ITB2/src/GammaFilter.kt
import itb2.filter.AbstractFilter
import itb2.image.Image
import java.lang.Math
import itb2.image.ImageFactory

class GammaFilter : AbstractFilter() {
    private fun calcGammaCorrection(x: Double, gamma: Double): Double {
        var x = x
        x = Math.round(256 * Math.pow(x / 255, gamma)).toDouble()
        return if (x == 256.0) 255.0 else x
    }

    override fun filter(input: Image): Image {
        val output: Image = ImageFactory.bytePrecision().gray(input.size)
        val gamma = properties.getDoubleProperty(GAMMA)
        for (col in 0 until input.width) {
            for (row in 0 until input.height) {
                var sum = 0.0
                for (chan in 0 until input.channelCount) sum += input.getValue(col, row, chan)
                output.setValue(col, row, calcGammaCorrection(sum / input.channelCount, gamma))
            }
        }
        return output
    }

    companion object {
        private const val GAMMA = "gamma"
    }

    init {
        properties.addDoubleProperty(GAMMA, 1.0)
    }
}